### PR TITLE
cmap, prov/rxm: Implement command queue

### PR DESCRIPTION
**Pull request by dmitrygladkov**
The command queue is used to notify main thread that owns device(domain) resources to perform actions based on commands from CM AV::CMAP.

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
This just refactors CMAP<->RxM::CQ interaction implemented in #4092 PR. This is done by means of CMD queue now.
If this approach is OK, the CMD queue is re-used to implement CM thread <-> main thread interaction.

File: include/ofi_util.h [hunk 1]
```diff
@@ -486,6 +486,75 @@ int ofi_get_src_addr(uint32_t addr_format,
...
+	assert(cmd_size >= sizeof(struct util_cmd_handle));
+	return queue->attr.cmd_alloc(cmd_size);
+}
+
+static inline void
+util_cmd_queue_free_cmd(struct util_cmd_queue *queue,
+			struct util_cmd_handle *handle)
+{
+	queue->attr.cmd_free(handle);
+}
```
Comment (shefty) [inline include/ofi_util.h]: This abstraction seems of questionable value.  The caller has to implement the alloc/free calls themselves.  The cmd_queue simply stores pointers to them, and then provides single line function wrappers around them.  The only real code it has is to queue/dequeue items into a double-linked list.

As an abstraction, we would be better off having a multi-threaded version of dlist.

File: include/ofi_util.h [hunk 2]
```diff
@@ -486,6 +486,75 @@ int ofi_get_src_addr(uint32_t addr_format,
...
+	return dlist_empty(&queue->queue);
+}
+
+/* Retrieve a command from the head of CMD queue */
+static inline struct util_cmd_handle *
+util_cmd_queue_read(struct util_cmd_queue *queue)
+{
+	struct util_cmd_handle *handle;
+
+	if (util_cmd_queue_empty(queue))
```
Comment (shefty) [inline include/ofi_util.h]: This check needs to be within the lock

File: include/ofi_util.h [hunk 3]
```diff
@@ -541,6 +610,7 @@ typedef int (*ofi_cmap_signal_func)(struct util_ep *ep, void *context,
 
 struct util_cmap_attr {
 	void 				*name;
+	struct util_cmd_queue		*cmd_queue;
```
Comment (shefty) [inline include/ofi_util.h]: This would be better of as a structure that is initialized, rather than a pointer that we also need to allocate.

File: include/ofi_util.h (hunk 1 above)
Comment (dmitrygladkov) [reply include/ofi_util.h]: Yes, I agree. do you mean to have lock-less version of dlist/slist?

File: include/ofi_util.h (hunk 2 above)
Comment (dmitrygladkov) [reply include/ofi_util.h]: agreed, here we have to protect an access

File: include/ofi_util.h (hunk 3 above)
Comment (dmitrygladkov) [reply include/ofi_util.h]: Ok, boss :) will do that

File: include/ofi_util.h (hunk 1 above)
Comment (shefty) [reply include/ofi_util.h]: The current version of dlist/slist are lockless.  I meant adding a lock-full? version.

Comment (dmitrygladkov) [reply include/ofi_util.h]: I'm not sure that this implementation is lockless, it rather lock-free, isn't it?

Comment (shefty) [reply include/ofi_util.h]: The current dlist implementation is lockless already.  The cmd queue is basically a wrapper around dlist, but adds locking.  I'm suggesting that a better abstraction than cmd queue is a thread-safe version of dlist -- one that adds locking.  Note that I'm not proposing that we add locking to dlist directly.  I'm suggesting a new version, say dlist_mt, that has a better name and locks around the dlist calls.

Comment (dmitrygladkov) [reply include/ofi_util.h]: Ok, got it.

**Comment by dmitrygladkov** [comment]
Starting from this patch, RxM/verbs works fine even if the library is built with --enable-debug. Before this patch, it detects data races with debug build.

**Comment by dmitrygladkov** [comment]
Also the abstraction allows to easily define new commands and add processing pf them for further needs

**Comment by shefty** [comment]
Btw, did you try writing a special error entry into the CQ that the main thread can use to process the deferred queue?  That should avoid needing a special command queue completely and fall through the existing code path without adding any new checks.

**Comment by dmitrygladkov** [comment]
> Btw, did you try writing a special error entry into the CQ that the main thread can use to process the deferred queue?

It would be great, but unfortunately, fi_cq_* api doesn't have fi_cq_write or fi_cq_write_error.

**Comment by shefty** [comment]
> It would be great, but unfortunately, fi_cq_* api doesn't have
> fi_cq_write or fi_cq_write_error.

We're writing directly into the rxm CQ, which is software based, from the rxm code.  We don't need to go through the external API.

**Comment by dmitrygladkov** [comment]
> We're writing directly into the rxm CQ

Do you mean the ofi_cq_write (from util/cq)? if yes, we need to protect CQ by locks in that case

File: prov/util/src/util_av.c [hunk 4]
```diff
@@ -1657,13 +1668,21 @@ struct util_cmap *ofi_cmap_alloc(struct util_ep *ep,
...
+	cmap->cmd_write = cmap->cmd_read = 0;
+
 	if (pthread_create(&cmap->event_handler_thread, 0,
 			   cmap->attr.event_handler, ep)) {
 		FI_WARN(ep->av->prov, FI_LOG_FABRIC,
 			"Unable to create msg_cm_listener_thread\n");
-		goto err3;
+		goto err4;
 	}
+	cmap->cm_thread_state = UTIL_CM_THREAD_OPEN;
```
Comment (a-ilango) [inline prov/util/src/util_av.c]: `cm_thread_state` is set at several places but it's not read anywhere.

File: prov/rxm/src/rxm.h [hunk 5]
```diff
@@ -619,9 +645,13 @@ rxm_ep_handle_unconnected(struct rxm_ep *rxm_ep, struct util_cmap_handle *handle
...
-		rxm_ep_progress_conn_deferred_list(
-					rxm_ep, container_of(handle, struct rxm_conn,
-							     handle));
+		struct rxm_conn *rxm_conn =
+			container_of(handle, struct rxm_conn, handle);
+		if (!dlist_empty(&rxm_conn->deferred_op_list)) {
+			rxm_ep_progress_conn_deferred_list(rxm_ep, rxm_conn);
+			if (dlist_empty(&rxm_conn->deferred_op_list))
+				dlist_remove(&rxm_conn->conn_deferred_entry);
+		}
```
Comment (a-ilango) [inline prov/rxm/src/rxm.h]: Should these go in a separate patch?

File: prov/rxm/src/rxm_conn.c [hunk 6]
```diff
@@ -136,59 +157,60 @@ static void rxm_send_queue_close(struct rxm_send_queue *send_queue)
...
 	struct rxm_ep *rxm_ep = container_of(handle->cmap->ep, struct rxm_ep, util_ep);
 	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
 
-	/* This handles case when saved_msg_ep wasn't closed */
-	if (rxm_conn->saved_msg_ep) {
-		rxm_ep_cleanup_posted_rx_list(rxm_ep, &rxm_conn->saved_posted_rx_list);
-		if (fi_close(&rxm_conn->saved_msg_ep->fid))
-			FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-				"Unable to close saved msg_ep\n");
+	if (handle->cmap == UTIL_CM_THREAD_OPEN) {
```
Comment (a-ilango) [inline prov/rxm/src/rxm_conn.c]: Should this be `handle->cmap->cm_thread_state`?

File: prov/util/src/util_av.c (hunk 4 above)
Comment (dmitrygladkov) [reply prov/util/src/util_av.c]: This is needed for CMAP_EXIT handling. Will think how to avoid this state

File: prov/rxm/src/rxm_conn.c (hunk 6 above)
Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Good catch. Thanks!

File: prov/rxm/src/rxm_conn.c [hunk 7]
```diff
@@ -287,28 +313,31 @@ rxm_msg_process_connreq(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
...
 err2:
 	ofi_cmap_del_handle(&rxm_conn->handle);
 err1:
 	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
 		"Rejecting incoming connection request\n");
-	if (fi_reject(rxm_ep->msg_pep, msg_info->handle, NULL, 0))
+	/*if (fi_reject(rxm_ep->msg_pep, msg_info->handle, NULL, 0))
 		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-				"Unable to reject incoming connection\n");
+		"Unable to reject incoming connection\n");*/
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Returned back and added proper handling of `ECONNREFUSED` event from error EQ when the remote side rejected a connection request

**Comment by dmitrygladkov** [comment]
@a-ilango I've updated the PR. Could you have a look again and provide new comments, please? I'll fix other comments soon.

File: include/ofi_queue.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+		}
+		root->head = next;
+		fastlock_release(&root->head_lock);
+
+		if (head == &root->marker) {
+			ofi_queue_push(head, root);
+			continue;
+		}
+		return head;
+	}
```
Comment (shefty) [inline include/ofi_queue.h]: This is very inefficient when the average queue depth is close to 1.  Why is this complexity needed, versus just wrapping slist with a single lock?

Comment (dmitrygladkov) [reply include/ofi_queue.h]: @shefty It seems that I found a real problem that cause issues with dlist_ts. I'm going to re-implement this PR in order to use dlist_ts. Are you OK with that?

**Comment by dmitrygladkov** [comment]
@a-ilango @shefty @vkrishna 
I've pushed a lot of new changes.
They fix a lot of problem related to RxM/verbs due to CM thread intervention to application thread resources.
Now it possible to run IMB, MPICH tests with MPICH/CH4 w/o any problems.

But... this PR revealed a problem with the TCP provider. The problem is fi_reject() called to reject an incoming connection request in case of bi-directional connection.
This leads that FI_SHUTDOWN is reported, but it hasn't be reported.

File: include/ofi_queue.h (hunk 8 above)
Comment (dmitrygladkov) [reply include/ofi_queue.h]: @shefty I remove the implementation of queue. Now CMD queue uses early implemented `dlist_ts`. But stil WIP, I need to address @a-ilango comments to split PR on separate commits

File: prov/rxm/src/rxm.h [hunk 9]
```diff
@@ -421,10 +421,50 @@ struct rxm_conn {
...
+
+	struct rxm_conn_close *close_data;
+};
+
+struct rxm_cmd_conn_free_data {
+	/* Must stay at top */
+	enum rxm_cmd_internal_type type; 
+
+	struct rxm_conn *rxm_conn;
 };
```
Comment (a-ilango) [inline prov/rxm/src/rxm.h]: Can we have a single common data struct? We can use an union inside to store different types of data.

File: prov/rxm/src/rxm_conn.c [hunk 10]
```diff
@@ -133,62 +128,110 @@ static void rxm_send_queue_close(struct rxm_send_queue *send_queue)
...
+	ssize_t ret;
+	struct rxm_cmd_connect_data connected_data = {
+		.type = RXM_CMD_CONNECTED,
+		.rxm_conn = rxm_conn,
+	};
 
-	if (!rxm_conn->saved_msg_ep)
+	cmd_connected = calloc(1, sizeof(struct util_cmap_cmd) +
+				  sizeof(struct rxm_cmd_connect_data));
+	if (!cmd_connected) {
```
Comment (a-ilango) [inline prov/rxm/src/rxm_conn.c]: This should never happen. We should assert(0) / return an error instead of working around.

File: prov/rxm/src/rxm.h [hunk 11]
```diff
@@ -421,10 +421,50 @@ struct rxm_conn {
...
+	struct fid_ep *msg_ep;
+	struct dlist_entry posted_rx_list;
+};
+
+enum rxm_cmd_internal_type {
+	RXM_CMD_CONNREQ_ACCEPT,
+	RXM_CMD_CONNECTED,
+	RXM_CMD_CONN_CLOSE,
+	RXM_CMD_CONN_FREE,
+};
```
Comment (a-ilango) [inline prov/rxm/src/rxm.h]: Can this be merged with cmap cmd type? There isn't anything RXM specific except for the data which is generic already.

Comment (dmitrygladkov) [reply prov/rxm/src/rxm.h]: Sure. I moved this into ofi_util.h

File: prov/rxm/src/rxm_conn.c (hunk 10 above)
Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Yes, this should never happen. But I think that using fallback mechanism (perform handling of the command here) is good  here.

File: prov/rxm/src/rxm.h (hunk 9 above)
Comment (dmitrygladkov) [reply prov/rxm/src/rxm.h]: Sure, nice point.

File: prov/rxm/src/rxm_conn.c [hunk 12]
```diff
@@ -287,28 +313,31 @@ rxm_msg_process_connreq(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
...
-				"Unable to accept incoming connection\n");
+	cmd_connreq = calloc(1, sizeof(struct util_cmap_cmd) +
+				sizeof(struct rxm_cmd_connreq_data));
+	if (!cmd_connreq)
 		goto err2;
-	}
+	cmd_connreq->type = UTIL_CMAP_CMD_INTERNAL;
+	connreq_data.rxm_conn = rxm_conn;
+	connreq_data.cm_data = cm_data;
+	memcpy(cmd_connreq->data, &connreq_data, sizeof(connreq_data));
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Done

File: prov/rxm/src/rxm_conn.c [hunk 13]
```diff
@@ -86,17 +87,37 @@ static int rxm_msg_ep_open(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
 		if (ret)
 			goto err;
 	}
-
+	assert(!rxm_conn->msg_ep);
```
[the remark this answers is not in the document]
Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Done, these changes were moved to the first commit

**Comment by dmitrygladkov** [comment]
@shefty @a-ilango This PR is waiting for your review :)

@vkrishna Can you check behavior of TCP provider? send calls contains check for `ENOTCONN`. if check for `ENOTOCONN` successful, it shutdowns connection.

**Comment by vkrishna** [comment]
@dmitrygladkov Yes tcp provider writes SHUTDOWN event to the eq whenever it finds out the socket is not connected. Do you see any issues with this behavior?

**Comment by dmitrygladkov** [comment]
> Do you see any issues with this behavior?

Yes, this cause error crash on RxM level. But if I remove the code that generates FI_SHUTDOWN, it causes a hang. Sorry, it's difficult to me to understand the actual reason of the problem due to small knowledge of TCP code

**Comment by dmitrygladkov** [comment]
@shefty @a-ilango Could you take a look, please? This PR fixes CM thread <-> application thread interactions that cause assertions (debug build) and sometimes crashes, because RxM touches RX/TX pools from different threads at the same time.

File: prov/rxm/src/rxm_cq.c [hunk 14]
```diff
@@ -1009,6 +1003,34 @@ static inline ssize_t rxm_ep_read_msg_cq(struct rxm_ep *rxm_ep)
...
+rxm_conn_handle_cmap_cmd_queue(struct rxm_ep *rxm_ep)
+{
+	ssize_t ret = 0;
+	struct util_cmap_cmd *cmd;
+
+	while (!dlist_ts_empty(&rxm_ep->util_ep.cmap->cmd_queue)) {
+		dlist_ts_pop_front(&rxm_ep->util_ep.cmap->cmd_queue,
+				   struct util_cmap_cmd, cmd, entry);
+		if (!cmd)
+			continue;
```
Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: Shouldn't this break out of the loop?

File: prov/rxm/src/rxm_cq.c [hunk 15]
```diff
@@ -1009,6 +1003,34 @@ static inline ssize_t rxm_ep_read_msg_cq(struct rxm_ep *rxm_ep)
...
+		rxm_ep->util_ep.cmap->cmd_read++;
```
Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: This should be serialized. May be use an atomic variable here?

File: prov/rxm/src/rxm_cq.c [hunk 16]
```diff
@@ -1009,6 +1003,34 @@ static inline ssize_t rxm_ep_read_msg_cq(struct rxm_ep *rxm_ep)
...
+		switch (cmd->type) {
+		case UTIL_CMAP_CMD_AV_UPD:
+			free(cmd);
+			rxm_ep->util_ep.cmap->av_upd_cmd = NULL;
```
Comment (a-ilango) [inline prov/rxm/src/rxm_cq.c]: We need to be holding cmap->lock when updating this variable. This can probably be pushed inside reprocess recv queue function where we acquire the lock.

File: prov/rxm/src/rxm.h [hunk 17]
```diff
@@ -421,10 +426,24 @@ struct rxm_conn {
...
+	struct rxm_conn *rxm_conn;
+
+	/* cmd-specific stuff */
+	union {
+		struct {
+			struct fi_info *msg_info;
+			struct rxm_cm_data cm_data;
+		} connreq_cmd;
+		struct {
+			struct rxm_conn_close *close_data;
```
Comment (a-ilango) [inline prov/rxm/src/rxm.h]: Can we avoid having to define a rxm_conn_close struct?

File: prov/rxm/src/rxm_conn.c [hunk 18]
```diff
@@ -199,18 +200,6 @@ static void rxm_conn_free(struct util_cmap_handle *handle)
...
-	if (!rxm_conn->saved_msg_ep)
-		return;
-	rxm_ep_cleanup_posted_rx_list(rxm_ep, &rxm_conn->saved_posted_rx_list);
-	/* Assuming fi_close also shuts down the connection gracefully if the
-	 * endpoint is in connected state */
-	if (fi_close(&rxm_conn->saved_msg_ep->fid))
-		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL, "Unable to close saved msg_ep\n");
-	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL, "Closed saved msg_ep\n");
-	rxm_conn->saved_msg_ep = NULL;
 }
```
Comment (a-ilango) [inline prov/rxm/src/rxm_conn.c]: This patch makes this function empty.

File: prov/rxm/src/rxm_conn.c [hunk 19]
```diff
@@ -350,6 +338,43 @@ static int rxm_conn_handle_notify(struct fi_eq_entry *eq_entry)
...
+	rxm_conn_wake_up_wait_obj(rxm_ep);
+}
+
+static void rxm_conn_handle_conn_refused(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn)
+{
+	if (rxm_conn->close_data) {
+		FI_DBG(&rxm_prov, FI_LOG_FABRIC,
+		       "Enqueue CONN_CLOSE req to CMD queue\n");
+		assert(rxm_conn->handle.state != CMAP_CONNREQ_SENT);
+		rxm_conn_handle_close(rxm_ep, rxm_conn);
```
Comment (a-ilango) [inline prov/rxm/src/rxm_conn.c]: It's not clear how rxm_conn->close_data would be set at this point.

File: prov/rxm/src/rxm_conn.c [hunk 20]
```diff
@@ -94,16 +91,36 @@ static int rxm_msg_ep_open(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
...
+	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL, "Closed msg_ep\n");
+	rxm_conn->msg_ep = NULL;
+}
+
+static void rxm_conn_wake_up_wait_obj(struct rxm_ep *rxm_ep)
+{
+	if (rxm_ep->util_ep.tx_cq->wait)
+		util_cq_signal(rxm_ep->util_ep.tx_cq);
+	if (rxm_ep->util_ep.tx_cntr && rxm_ep->util_ep.tx_cntr->wait)
+		util_cntr_signal(rxm_ep->util_ep.tx_cntr);
```
Comment (vkrishna) [inline prov/rxm/src/rxm_conn.c]: Don't we need to signal rx_cq and rx_cntr, in case if they are different?

Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Good point, we need it

File: prov/rxm/src/rxm_conn.c (hunk 18 above)
Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Yes, right, this is not needed anymore. Let's remove this fp

File: prov/rxm/src/rxm_conn.c [hunk 21]
```diff
@@ -360,9 +385,10 @@ static void rxm_conn_handle_eq_err(struct rxm_ep *rxm_ep, ssize_t rd)
 	}
 	OFI_EQ_READERR(&rxm_prov, FI_LOG_FABRIC, rxm_ep->msg_eq, rd, err_entry);
 	if (err_entry.err == ECONNREFUSED) {
-		FI_DBG(&rxm_prov, FI_LOG_FABRIC, "Connection refused\n");
-		ofi_cmap_process_reject(rxm_ep->util_ep.cmap,
-					err_entry.fid->context);
+		struct rxm_conn *rxm_conn =
+			container_of(err_entry.fid->context,
+				     struct rxm_conn, handle);
```
Comment (vkrishna) [inline prov/rxm/src/rxm_conn.c]: Here rxm_conn may have a handle with values updated with a valid connection info, in case of simultaneous connections. It may not have info about refused msg endpoint. Because of that you may end up closing the wrong msg ep while processing UTIL_CMAP_CMD_CONN_CLOSE in cmd queue. This is the reason why you are getting FI_SHUTDOWN from tcp provider in certain cases.

Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: Yes, right. A handle can contain a new connection info. We use `close_data` field from `rxm_conn` that carries MSG EP that should be closed

File: prov/rxm/src/rxm_conn.c (hunk 19 above)
Comment (dmitrygladkov) [reply prov/rxm/src/rxm_conn.c]: rxm_conn->close_data is allocated when handling simultaneous connection request from passive side of connection (that has lower address). When ECONNREFUSED error received that notifies us about REJECT event, we will close MSG EP that contains in close_data.

File: prov/rxm/src/rxm.h (hunk 17 above)
Comment (dmitrygladkov) [reply prov/rxm/src/rxm.h]: Yes, we can, but we need to carry MSG_EP and posted_rx_list. What do you think about this approach?

Comment (a-ilango) [reply prov/rxm/src/rxm.h]: I have a PR in the works that eliminates the responsibility to maintaining posted_rx_list. IIUC, the list is needed to track posted receive buffers only to keep the util buffer pool happy in debug mode.

File: prov/rxm/src/rxm_conn.c (hunk 21 above)
Comment (a-ilango) [reply prov/rxm/src/rxm_conn.c]: At the time of receiving FI_REJECT, `rxm_conn->msg_ep` would already be overwritten with new  `msg_ep` that was created by accepting connection request from remote. This `msg_ep` should not be closed. En-queuing `UTIL_CMAP_CMD_CONN_CLOSE` command should be done as part of `rxm_conn_close`.

File: prov/rxm/src/rxm_conn.c [hunk 22]
```diff
@@ -136,59 +153,82 @@ static void rxm_send_queue_close(struct rxm_send_queue *send_queue)
...
-	dlist_splice_tail(&rxm_conn->saved_posted_rx_list,
+
+	rxm_conn->close_data = calloc(1, sizeof(*rxm_conn->close_data));
+	if (!rxm_conn->close_data) {
+		assert(0);
+		rxm_conn_close_msg_ep(rxm_ep, rxm_conn);
+		return;
+	}
+
+	rxm_conn->close_data->msg_ep = rxm_conn->msg_ep;
```
Comment (vkrishna) [inline prov/rxm/src/rxm_conn.c]: ofi_cmap_process_connreq() in util_av.c seems to be calling this rxm_conn_close()  with the rxm_conn->msg_ep is for which the connection is accepted. So when the sent out connreq is refused by peer we are closing this msg_ep with valid connection. Am I missing something here?

File: prov/rxm/src/rxm_conn.c (hunk 21 above)
Comment (a-ilango) [reply prov/rxm/src/rxm_conn.c]: Sorry. On a second glance, I think we're okay here.

File: prov/rxm/src/rxm.h (hunk 17 above)
Comment (dmitrygladkov) [reply prov/rxm/src/rxm.h]: Thanks, Arun! I've re-based this PR on top of your changes (master branch).

Comment (dmitrygladkov) [reply prov/rxm/src/rxm.h]: rxm_close_data has been removed as no longer needed

File: prov/rxm/src/rxm_conn.c (hunk 20 above)
Comment (a-ilango) [reply prov/rxm/src/rxm_conn.c]: Was this comment addressed?

File: prov/rxm/src/rxm_conn.c [hunk 23]
```diff
@@ -143,44 +159,61 @@ static void rxm_conn_close(struct util_cmap_handle *handle)
...
+rxm_conn_process_connected(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn)
+{
+	struct util_cmap_cmd *cmd;
+	struct rxm_cmap_cmd_data *cmd_data;
+
+	cmd = calloc(1, sizeof(struct util_cmap_cmd) +
+			sizeof(struct rxm_cmap_cmd_data));
+	if (!cmd) {
+		if (!rxm_ep->srx_ctx) {
+			(void) rxm_ep_prepost_buf(rxm_ep, rxm_conn->msg_ep);
```
Comment (a-ilango) [inline prov/rxm/src/rxm_conn.c]: Can we add an assert for memory allocation failure? I think we need to handle memory allocation errors instead of ignoring them. It can be done in a follow-up patch.

File: prov/rxm/src/rxm_conn.c (hunk 18 above)
Comment (a-ilango) [reply prov/rxm/src/rxm_conn.c]: The empty function is still there.

File: prov/rxm/src/rxm_conn.c (hunk 23 above)
Comment (shefty) [reply prov/rxm/src/rxm_conn.c]: Memory allocation failures need to be handled.  This is an error that klocworks will report and must be handled when integrated libfabric into software products.

**Comment by dmitrygladkov** [comment]
Thanks for reviewing this. But unfortunately, this shows a very slow startup. I decided to close this PR
